### PR TITLE
Develop2 branch

### PR DESCRIPTION
**Pull request by dmmurill0**
- Refactored db interaction out of keyword_analyzer.py, specifically in the Analyzer class and analyze function, so that db interaction is done through views.py
- Added NLTK stopwords to dev_seed.json, creating a new file dev_seed_with_stopwords.json
- Created 6 unique tests to ensure expected functionality among entire codebase
- Added coverage report to identify percentage of code touched by tests

### CODE REVIEW

File: cold_apply/models.py [hunk 1]
```diff
@@ -195,10 +195,16 @@ class Meta:
...
+    CATEGORY_CHOICES = [
+        ('nltk english stopwords', 'NLTK English Stopwords'),
+        ('low value', 'Low Value'),
+        ('industry protected', 'Industry Protected'),
+    ]
     job = models.ForeignKey(Job, on_delete=models.CASCADE)
-    unigram = models.CharField(max_length=200)
-    bigram = models.CharField(max_length=200)
-    trigram = models.CharField(max_length=200)
+    category = models.CharField(max_length=50, choices=CATEGORY_CHOICES)
```
Comment (janton42) [inline cold_apply/models.py]: @dmmurill0 - why were the max_length values changed?

File: cold_apply/models.py [hunk 2]
```diff
@@ -195,10 +195,16 @@ class Meta:
 # Results of keyword analysis algorithm. There is one per job, and it is overwritten whenever a new keyword analysis
 # is run.
 class KeywordAnalysis(models.Model):
+    CATEGORY_CHOICES = [
+        ('nltk english stopwords', 'NLTK English Stopwords'),
```
Comment (janton42) [inline cold_apply/models.py]: @dmmurill0 - are these categories needed here? This class is for an analysis of a particular job post, so it only needs to hold the different N-grams resulting from the keyword analyzer function. The category choice will be made elsewhere, and I don't think there's a need to label the results of an analysis with them in any case.

File: cold_apply/models.py [hunk 3]
```diff
@@ -227,9 +233,9 @@ class Meta:
 
 class BulletKeyword(models.Model):
     bullet = models.ForeignKey(Bullet, on_delete=models.CASCADE)
-    unigram = models.CharField(max_length=200)
-    bigram = models.CharField(max_length=200)
-    trigram = models.CharField(max_length=200)
+    unigram = models.CharField(max_length=5000)
+    bigram = models.CharField(max_length=5000)
+    trigram = models.CharField(max_length=5000)
```
Comment (janton42) [inline cold_apply/models.py]: @dmmurill0 - why were these max_length values changed?

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 4]
```diff
@@ -2,163 +2,105 @@
...
         text = ''.join([c for c in self.input_text if c not in string.punctuation])
-        # Split description into words
         tokens = text.split()
-        # Remove stopwords
         words = [word for word in tokens if word not in self.stopwords]
-        # return cleaned data as a list of words
         return words
 
     def find_keywords(self, lower_bound, upper_bound):
-        """
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:75]: @dmmurill0 - why did you remove this docstring?

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 5]
```diff
@@ -2,163 +2,105 @@
...
+        )
+        # Use cleaned words for TF-IDF
+        cleaned_text = ' '.join(self.cleaned_words)
+        X_tfidf = vectorizer.fit_transform([cleaned_text])
         feature_names = vectorizer.get_feature_names_out()
         X_tfidf_df = pd.DataFrame(X_tfidf.toarray())
         X_tfidf_df.columns = feature_names
         X_tfidf_df.sort_values(by=X_tfidf_df.index[0], axis=1, inplace=True, ascending=False)
         top_twenty = X_tfidf_df.iloc[:, :20].columns.tolist()
-        # Alphabetize multi-word ngrams
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:90]: @dmmurill0 - why did you remove this comment?

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 6]
```diff
@@ -2,163 +2,105 @@
...
 # Writes results to database
 def hook_after_jd_analysis(task, job_id: int):
-    data = json.loads(task)
-    serialized = list()
-    for line in data:
-        data[line]['job'] = job_id
-        formatted = dict(
-            model='cold_apply.keywordanalysis',
-            fields=data[line]
+    if isinstance(task, dict):
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: @dmmurill0 - I think I'm following the logic here, but am unclear why this check is needed. Isn't this logic handled in the view that calls this function?

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 7]
```diff
@@ -2,163 +2,105 @@
...
+        data = task
+    else:
+        data = json.loads(task)
+    job = Job.objects.get(id=job_id)
+    for category in data:
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: @dmmurill0 - this category filtering feels misplaced. The categories should be linked in the DB to lists of stopwords, not to n-grams that resulted from an analysis

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 8]
```diff
@@ -2,163 +2,105 @@
...
-    serialized = list()
-    for line in data:
-        data[line]['bullet'] = bullet_id
-        formatted = dict(
-            model='cold_apply.bulletkeyword',
-            fields=data[line]
-        )
-        serialized.append(formatted)
+    serialized = []
+    for category in data:
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: @dmmurill0 - this also appears to be misplaced logic because of the category labeling in the DB, which shouldn't be on the n-gram results.

File: cold_apply/views.py [hunk 9]
```diff
@@ -347,9 +348,20 @@ class JobDetailView(LoginRequiredMixin, DetailView):
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         if KeywordAnalysis.objects.filter(job=self.object).exists():
+            print(f"DEBUG: Recalling existing keyword analysis for job {self.object.id}")
             context["keywords"] = KeywordAnalysis.objects.filter(job=self.object)
         else:
-            analysis = analyze(self.object.description)
+            print(f"DEBUG: Running new keyword analysis for job {self.object.id}")
```
Comment (janton42) [inline cold_apply/views.py:354]: @dmmurill0 - this looks good, and is related to the comments in models.py. The logic here should work to grab the desired stopwords by category, then group them into one list, which is sent to the analyze function along with the JD. This is the only place categories is relevant in the overall logic, because it is part of the INPUT for the keyword analysis. The OUTPUT, which is the n-grams, does not need to have the category labels.

File: hl_main/settings/dev.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 INTERNAL_IPS = ["127.0.0.1"]
 MIDDLEWARE += [
     "debug_toolbar.middleware.DebugToolbarMiddleware",
-    "django_browser_reload.middleware.BrowserReloadMiddleware",
```
Comment (janton42) [inline hl_main/settings/dev.py:15]: @dmmurill0 - why was this removed? It's a fairly common best practice to use trailing commas at the end of lists, and we definitely do that in this project.

**Review by janton42 — CHANGES_REQUESTED** [review]
@dmmurill0 - looks pretty good, but I have a questions before merging, which I entered as comments where relevant.

File: cold_apply/models.py (hunk 1 above)
Comment (dmmurill0) [reply cold_apply/models.py]: I had changed the values because my previous logic inside keyword_analyzer was requiring analyze() to persist data to the db, creating large strings/lists of words to be stored in models. Reverting the values back was an oversight once I refactored.

File: cold_apply/models.py (hunk 2 above)
Comment (dmmurill0) [reply cold_apply/models.py]: It was my way of trying to make sure the categories didn't get lost, unaware that views.py handles categorization.

File: cold_apply/models.py (hunk 3 above)
Comment (dmmurill0) [reply cold_apply/models.py]: This was a workaround that I forgot to revert back once I refactored. I had to increase the length so that BulletKeyword class would accept the values for each of the model fields.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 4 above)
Comment (dmmurill0) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:75]: This may have been a result of my first use of ChatGPT's codex extension. It was making all kinds of edits before I realized it was doing more than just having a conversation.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 5 above)
Comment (dmmurill0) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:90]: Again, initial use of codex extension.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 6 above)
Comment (dmmurill0) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: This check was when I was transitioning from a json string and instead using a dictionary, so it was to make sure the hook knew which one to accept. But after my refactor was finalized with the dictionary there was no longer a need to check it. In short, it was an oversight that I didn't remove after refactor.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 7 above)
Comment (dmmurill0) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: Views already handles the categorization, so i see this edit in hooks as redundant. I am realizing a fundamental issue is that I've been trying to get one file to do it all, rather than understanding that each file serves its own purpose: models -> db schema; views -> orchestrates the flow; keyword_analyzer -> pure logic (with the exception of the two hooks that write to db).

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 8 above)
Comment (dmmurill0) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py]: Same as above, I misunderstood how views was already handling the categorization, requiring hooks to only write n-gram results to db.

File: cold_apply/views.py (hunk 9 above)
Comment (dmmurill0) [reply cold_apply/views.py:354]: I'm finally seeing how and why the pieces fit together.

File: hl_main/settings/dev.py (hunk 10 above)
Comment (dmmurill0) [reply hl_main/settings/dev.py:15]: I don't remember if I had opened up the file to learn something and inadvertently took it out, or if this was something the codex did. But either way I don't recall why that file was ever accessed. Good to know as far as common best practices.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 4 above)
Comment (janton42) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:75]: Small thing, but please fix it. The comments are inconsistent, but we don't want to remove any.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py (hunk 5 above)
Comment (janton42) [reply cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:90]: This is another very small item that needs to be fixed before merge.

File: hl_main/settings/dev.py (hunk 10 above)
Comment (janton42) [reply hl_main/settings/dev.py:15]: Last remaining small item from the original change request - please fix.

File: cold_apply/models.py [hunk 11]
```diff
@@ -197,8 +197,8 @@ class Meta:
 class KeywordAnalysis(models.Model):
     job = models.ForeignKey(Job, on_delete=models.CASCADE)
     unigram = models.CharField(max_length=200)
-    bigram = models.CharField(max_length=200)
```
Comment (janton42) [inline cold_apply/models.py:200]: @dmmurill0 - 200 was not chosen entirely at random, but it's way more than enough space; 200 characters is something like 28 words, when our expected data is no more than 3 words for trigrams. Why did you choose other values? We discussed anything in the thousands being unreasonable, but ANY change to the size of the existing database is extremely dangerous, and must be done exceptionally cautiously and intentionally. 

200 is more than enough - what is your rationale for setting it to 500 characters here? That's ~75-125 words, when we only need space for 2.

File: cold_apply/models.py [hunk 12]
```diff
@@ -197,8 +197,8 @@ class Meta:
...
-    trigram = models.CharField(max_length=200)
+    bigram = models.CharField(max_length=500)
+    trigram = models.CharField(max_length=800)
```
Comment (janton42) [inline cold_apply/models.py:201]: What is your rationale for setting this to 800 characters? That's space for 150-200 words, but we only need space for 3.

File: cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py [hunk 13]
```diff
@@ -2,163 +2,90 @@
...
+    a = Analyzer(job_description, stopwords=stopwords or [])
+    return {
+        "unigram": a.find_keywords(1, 1),
+        "bigram":  a.find_keywords(2, 2),
+        "trigram": a.find_keywords(3, 3),
+        }
+
+# Hooks allowed DB access
+from django.core import serializers
+from cold_apply.models import KeywordAnalysis, Job
```
Comment (janton42) [inline cold_apply/static/scripts/keyword_analyzer/keyword_analyzer.py:61]: @dmmurill0 - this still isn't quite right. There should be NO connections directly between ANY FILE in the static/scripts/ directory or its subdirectories and anything Django-related. So, there is definitely no need to import anything from cold_apply.models or django.core.

ALL of the interactions between ANY file in a static/ directory or subdirectory should be mediated by the views.py file. If you find yourself adding an import statement from anything Django-related into a static script file, stop, because that connection should not be made in the static file.

Instead, the views.py file should be the link between the models.py and the scripts in the static/ directory. All the import statements from models.py or Django-related stuff, like serializers, should be done in the views.py. Functions from static files should also be imported there. 

So, the logic is like this:
Models only define database tables, scripts only define functions and calculations, views tie it all together for processing of data and presentation to a user in a web browser.

**Review by janton42 — CHANGES_REQUESTED** [review]
A few small things remain from the first review (all minor punctuation and comments). There is still some confusion around where and how data is being processed. That's total normal - it's all very confusing. 

Hopefully my comments are helpful, but let me know if you have any questions.
